frontend/components/ClientComponent.tsx

```diff
@@ -0,0 +1,22 @@
+import React from 'react';
+
+export interface ClientComponentProps<T> {
+    f: () => Promise<T>;
+    children: React.SFC<{ data: T | undefined }>;
+}
+
+export class ClientComponent<T> extends React.Component<
+    ClientComponentProps<T>,
+    { data: T | undefined }
+> {
+    public state = { data: undefined };
```

Comment:
Can you get away with not declaring the `data` property here, or does `tsc` throw a wobbly?

Reply:
Unfortunately that results in state being undefined. 

I'm not 100% sure that we should feed it undefined, but it feels simplest.

Reply:
(As in that we use data = undefined to mean both no response or failed response.

Reply:
So if you changed it to:
```js
public state = {};
```
TypeScript would interpret that as `state` being `undefined`?

Reply:
I think I might just have been over cautious, will take a deeper look at.